src/dns-test.js

```diff
@@ -317,7 +389,36 @@ async function runMeasurement() {
     let payload = {reason: STUDY_MEASUREMENT_COMPLETED};
     payload.dnsData = dnsData;
     payload.dnsAttempts = dnsAttempts;
+
+    // Run the fetch test one more time before submitting our measurements
+    await fetchTest();
+
+    // If we have passed the XHR test a second time, submit our measurements
     sendTelemetry(payload);
 }
 
-runMeasurement();
+/**
+ * Entry point for our addon.
+ */
+async function main() {
+    // If we can't upload telemetry. don't run the addon
+    let canUpload = await browser.telemetry.canUpload();
+    if (!canUpload) {
+        throw new Error(STUDY_ERROR_TELEMETRY_CANT_UPLOAD);
+    }
+
+    // Use the captive portal API to determine if we have Internet connectivity.
+    // If we already have connectivity, run the measurement.
+    // If not, wait until we get connectivity to run it.
+    let captiveStatus = await browser.captivePortal.getState();
```

Comment:
This method can throw if the captivePortal API is unavailable* (e.g. because one has disabled the feature via prefs)? Wouldn't you want to run the measurement anyway? Or is it preferred to exit early?

\* see https://searchfox.org/mozilla-central/rev/05cea55db68db7a3c8531cadaa169e46176e3983/toolkit/components/extensions/parent/ext-captivePortal.js#49-53,56-58

Reply:
Oh interesting, I haven't thought about that. Let me get back to you.

Reply:
We decided to exit the measurement and queue some telemetry to report the error. Addressed by 104ad02